File: csharp-workshop/Part2.LightMocks/Service/OrderResult.cs
using Part2.LightMocks.Domain;

namespace Part2.LightMocks.Service;

public abstract record OrderResult
{
    public sealed record Success(Order Order, ShippingConfirmation ShippingConfirmation) : OrderResult;
    public sealed record Failure(Order Order, string Reason) : OrderResult;
}

File: typescript-workshop/part-3-mocks-clean/src/service/OrderResult.ts
import { Order } from '../domain/Order';
import { ShippingConfirmation } from '../domain/ShippingConfirmation';

export type OrderResult =
  | { readonly kind: 'success'; readonly order: Order; readonly shippingConfirmation: ShippingConfirmation }
  | { readonly kind: 'failure'; readonly order: Order; readonly reason: string };

export const OrderResult = {
  success: (order: Order, shippingConfirmation: ShippingConfirmation): OrderResult =>
    ({ kind: 'success', order, shippingConfirmation }),
  failure: (order: Order, reason: string): OrderResult =>
    ({ kind: 'failure', order, reason }),
};
